crates/ra_assists/src/lib.rs

```diff
@@ -63,7 +64,7 @@ where
 ///
 /// Assists are returned in the "resolved" state, that is with edit fully
 /// computed.
-pub fn assists<H>(db: &H, range: FileRange) -> Vec<(AssistLabel, AssistAction)>
+pub fn assists<H>(db: &H, range: FileRange) -> Vec<(AssistLabel, Vec<AssistAction>)>
```

Comment:
I thin the return type should look like this:

```rust
pub fn assists<H>(db: &H, range: FileRange) -> 
  (
    Vec<WithLabel<
      Either<
        AssistId, 
        Vec<WithLabel<AssistId>>
      >
    >>, 
    FxHashMap<AssistId, AssistAction>
  ) 

{ ... }

struct WithLabel<T> {
  label: String,
  value: T,
}

```

but we probably should introduce some named types to avoid `>>>>>`.

In other words, with lhs we should be able to draw a depth-one tree

```rust
do foo (id1)
do baz (id2)
  with spam (id3)
  with eggs (id4)
do quux (id5)
```

and with rhs we shoudl be able to map each id to an action!

Reply:
I've tried this approach btw and it's not that straightforward, I think:

1. Currently `AssistAction` has an optional label that should be set in the builder, so we cannot really return `Vec<WithLabel<AssistId>>` currently.
Forcing the users to input the label every time means changing all current assists and duplicating the Assist label for a single `AsssistAction`.

1. There's no `FxHashMap` crate in `ra_assists` and it feels like we don't really need it that much: it looks like we should rather return `Vec<Assist>` where `Assist` will be something like
```rust
struct Assist {
  label: String,
  id: AssistId,
  action_data: Either<AssistAction, Vec<AssistAction>>,
}
```

or leave it as it is currently (with `alternative_actions`).
Former appoach will also require some additional changes in tests again, since now they are relying on a single action to be present.

So it feels like maximum I can do here is to add some wrapper named struct to hold all data returned and a few methods to it.